tket/src/Transformations/BasicOptimisation.cpp

```diff
@@ -206,33 +224,33 @@ static bool commute_singles_to_front(Circuit &circ) {
     while (!is_initial_q_type(circ.get_OpType_from_Vertex(current_v))) {
       const Op_ptr curr_op = circ.get_Op_ptr_from_Vertex(current_v);
       // if current vertex is a multiqubit gate
-      if (circ.n_in_edges_of_type(current_v, EdgeType::Quantum) > 1 &&
-          curr_op->get_desc().is_gate()) {
-        // need gate check to access commuting_basis
-        const std::pair<port_t, port_t> ports = circ.get_ports(current_e);
-      check_prev_commutes:
-        const Op_ptr prev_op = circ.get_Op_ptr_from_Vertex(prev_v);
-        // if previous vertex is single qubit gate
-        if (prev_op->get_desc().is_gate() &&
-            circ.n_in_edges_of_type(prev_v, EdgeType::Quantum) == 1) {
-          const std::optional<Pauli> prev_colour =
-              circ.commuting_basis(prev_v, PortType::Target, ports.second);
-
-          if (circ.commutes_with_basis(
-                  current_v, prev_colour, PortType::Source, ports.first)) {
-            // subsequent op on qubit path is a single qubit gate
-            // and commutes with current multi qubit gate
-            success = true;
-            circ.remove_vertex(
-                prev_v, Circuit::GraphRewiring::Yes,
-                Circuit::VertexDeletion::No);
-            Edge rewire_edge = circ.get_nth_in_edge(current_v, ports.first);
-            circ.rewire(prev_v, {rewire_edge}, {EdgeType::Quantum});
-            current_e = circ.get_nth_out_edge(current_v, ports.first);
-            prev_v = circ.target(current_e);
-            // check if new previous gate can be commuted through too
-            goto check_prev_commutes;
+      if (circ.n_in_edges_of_type(current_v, EdgeType::Quantum) > 1) {
+        while (circ.n_in_edges_of_type(prev_v, EdgeType::Quantum) == 1 &&
+               ends_commute(circ, current_e)) {
+          // subsequent op on qubit path is a single qubit gate
+          // and commutes with current multi qubit gate
+          success = true;
+          EdgeVec rewire_edges;
+          op_signature_t edge_types;
+          for (const Edge &e : circ.get_in_edges(prev_v)) {
+            EdgeType type = circ.get_edgetype(e);
+            Edge boundary_edge;
+            if (type == EdgeType::Quantum) {
+              boundary_edge = circ.get_last_edge(current_v, current_e);
+            } else {
+              // There should not be any Classical edges for conditional ops
+              TKET_ASSERT(type == EdgeType::Boolean);
```

Comment:
I'm still a bit uneasy about this -- even though you added a test with conditional measurements. (Are we OK there because they only have one quantum wire?) I don't think we can really guarantee this, so would prefer an exception be thrown rather than an assert.

Reply:
Of course I can replace this with an exception. But this is really something that should be documented and we should have a definite answer to. More generally a docs page about how classical wires and conditions work exactly would be very useful.

I have to say that personally I keep getting surprised with conditional wires -- hence also this bug for example.

Reply:
Done :)